bug fix for geo indexing and display

lmgonzales opened the pull request:
Fixing two main issues with this pull request:
1. Trap for errors in geographic coordinates so that Solr does not run out of memory.
2. Fix a bug to correct the display of rectangles crossing the dateline.  

See also https://vufind.org/jira/browse/VUFIND-1231

demiankatz commented:
Thanks, @lmgonzales! Is there any chance you could include some test records that trigger these warnings so I can make sure I'm setting everything up correctly when I port the code forward to the latest version? I also plan to look into converting these to use SolrMarc's built in logger object instead of System.out.println for greater flexibility of reporting.

lmgonzales commented:
Sure thing @demiankatz. I will put together a test set of records this week. Should I include the test files in this PR or send them to you directly?  Thanks for converting the System.out.println into the SolrMarc's logger object.

demiankatz commented:
Perhaps the best thing is to add the test records to the end of the existing file of geographic test data (https://github.com/vufind-org/vufind/blob/master/tests/data/geo.mrc) -- this way we'll have a regression test for the future -- if something breaks in the indexing, it will make test builds fail and allow us to investigate.

lmgonzales commented:
Sounds good. I'll add the new test records to that file.

demiankatz commented on import/index_java/src/org/solrmarc/index/VuFindIndexer.java:
```diff
@@ -1624,19 +1631,136 @@ protected Double coordinateToDecimal(String coordinateStr) {
     protected boolean validateCoordinates(Double west, Double east, Double north, Double south) {
-        if (west == null || east == null || north == null || south == null) {
-            return false;
+        String validLines = "";
```
Is there a specific reason that you chose String types here rather than booleans?

lmgonzales replied:
If we can use booleans, then yes, let's do that instead. I'm all for refactoring to make better code. Thank you for the suggestions

demiankatz commented on import/index_java/src/org/solrmarc/index/VuFindIndexer.java:
```diff
@@ -1624,19 +1631,136 @@ protected Double coordinateToDecimal(String coordinateStr) {
+            validLines = "true";
+        } else {
+            validLines = "false";
+            System.out.println("*** ERROR: Coordinates form a line at the pole...DD coords: "+west+" "+east+" "+north+" "+south);
```
Is there a reason to put the output here, rather than inside the validateLines function?

demiankatz replied:
(I'm just thinking that if we move the output inside the validation functions, it will make this long calling function smaller and more readable).

lmgonzales replied:
Yes, good point. I agree that would be a much better approach.

demiankatz replied:
Great, thanks! I might have some time to work on this tomorrow, unless you want to make any changes in the meantime. If I do end up changing anything, I'll push my changes here and get your approval before moving forward.

lmgonzales commented:
Thanks Demian. I’ll see if I have some time to work on this today.



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Thursday, May 18, 2017 9:22 AM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] bug fix for geo indexing and display
(#964)



*@demiankatz* commented on this pull request.
------------------------------

In import/index_java/src/org/solrmarc/index/VuFindIndexer.java
<https://github.com/vufind-org/vufind/pull/964#discussion_r117292999>:

> @@ -1624,19 +1631,136 @@ protected Double coordinateToDecimal(String coordinateStr) {

      * @return boolean

      */

     protected boolean validateCoordinates(Double west, Double east,
Double north, Double south) {

-        if (west == null || east == null || north == null || south == null) {

-            return false;

+        String validLines = "";

+        String validValues = "";

+        String validExtent = "";

+        String validNorthSouth = "";

+        String validCoordDist = "";

+

+        // Validate lines

+        if (validateLines(west, east, north, south)) {

+            validLines = "true";

+        } else {

+            validLines = "false";

+            System.out.println("*** ERROR: Coordinates form a line at
the pole...DD coords: "+west+" "+east+" "+north+" "+south);

Great, thanks! I might have some time to work on this tomorrow, unless you
want to make any changes in the meantime. If I do end up changing anything,
I'll push my changes here and get your approval before moving forward.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/964#discussion_r117292999>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0URQvZ6Kbfmc2yh4vPZRBu7bzSkLkks5r7HAOgaJpZM4NV6cF>
.[image: Image removed by sender.]

demiankatz commented:
Thanks, @lmgonzales. Unfortunately, it doesn't look like I'm going to get to this before next week (and next week is looking pretty busy already). However, if you're not able to get to it yourself, let me know and I'll try to schedule it as soon as I can!

demiankatz commented:
@lmgonzales, I actually did find a couple of minutes of free time today, so I went ahead and replaced your System.out.println calls with logger.error calls, which should make the messages integrate a little more nicely with other output from the indexer. You could also change this to logger.info or logger.warn if you don't think we should consider these to be errors.

I notice that with the test data you sent, I'm getting null pointer exceptions in addition to the error messages you are outputting. Is this to be expected, or do we need to figure out a way to make this fail more gracefully?

lmgonzales commented:
Thanks Demian. I think we want to flag these as errors since the
coordinates won’t be indexed. That way whomever is loading the file can be
made aware that they have problems with the coordinates in their dataset.



As for the test data set, no, it shouldn’t be throwing null pointer
exceptions. Let me take a look at it and get back to you on it. I should
have some time this afternoon to look it over.



Thanks again for your help,

Leila



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Friday, May 19, 2017 9:01 AM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] bug fix for geo indexing and display
(#964)



@lmgonzales <https://github.com/lmgonzales>, I actually did find a couple
of minutes of free time today, so I went ahead and replaced your
System.out.println calls with logger.error calls, which should make the
messages integrate a little more nicely with other output from the indexer.
You could also change this to logger.info or logger.warn if you don't think
we should consider these to be errors.

I notice that with the test data you sent, I'm getting null pointer
exceptions in addition to the error messages you are outputting. Is this to
be expected, or do we need to figure out a way to make this fail more
gracefully?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/964#issuecomment-302742894>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0UbAksvooBQKfo1Mw55ZJotRy-bj7ks5r7bytgaJpZM4NV6cF>
.[image: Image removed by sender.]

lmgonzales commented:
@demiankatz, 
I updated the code to catch the null pointer exception errors from the geo.mrc test set. The geo.mrc test file has some very poorly formed coordinates (on purpose), so I've reworked the routines to catch them all. The previous code missed a few of the odd coordinates, but the error messages now make it more clear as to what's amiss with the coordinates if they fail the indexing routine. 

I also refactored the validation routines per your suggestions about using booleans and putting the error messages in the validation subroutines. 

Also, I converted all the System.out.println calls to use the logger.error method. 

Please let me know if you find anything that needs fixing, or if you see areas where the code could be improved further. Thanks!

demiankatz commented on import/index_java/src/org/solrmarc/index/VuFindIndexer.java:
```diff
@@ -1444,35 +1444,59 @@ public static String getDeweySortableByType(
-                    e = d;
-                    g = f;
+                // Initialize coordinate variables
+                String d_coord = "";
```
Should these be initialized to null rather than the empty string? I see you are checking for null below, but empty string is a non-null value.

demiankatz replied:
(Note that if you follow up on my refactoring suggestions in the next comment, this issue goes away by itself).

lmgonzales replied:
Good catch. Yes, I should have initialized them to null, and better yet, I think your refactoring suggestions below are the way to go.

demiankatz commented on import/index_java/src/org/solrmarc/index/VuFindIndexer.java:
```diff
@@ -1444,35 +1444,59 @@ public static String getDeweySortableByType(
+                String f_coord = "";
+                String g_coord = "";
+
+                if (df.getSubfield('d') != null) {
```
It might be slightly more efficient to replace this with something like:

<pre>
Subfield d_subfield = df.getSubfield('d');
if (d_subfield != null) {
    d_coord = d_subfield.getData();
}
</pre>

(to avoid doing the subfield lookup twice).

It might simplify the code even further to use a HashMap<Character, String> to map subfield to data value -- then you could do this retrieval in a loop instead of copying and pasting the logic multiple times. You could do something like:

<pre>
HashMap<Character, String> coords = new HashMap<Character, String>();
for (char code = 'd'; code <= 'g'; code++) {
    Subfield subfield = df.getSubfield(code);
    if (subfield != null) {
        coords.put(code, subfield.getData());
    }
}
</pre>

...instead of having four separate retrievals/validations/variables.

Then later in the code, instead of doing a null check, you can do `coords.containsKey('a')` to check if a value is set, and `coords.get('a')` to fetch the retrieved value.

I notice that there is a lot of logic duplicated between getAllCoordinates(), getDisplayCoordinates() and getPointCoordinates(). By using this HashMap approach, you might be able to deduplicate that logic, creating a method that takes in a data field and spits out a HashMap of validated/normalized coordinates, thus making all of those methods shorter and more readable.

lmgonzales replied:
Okay, that sounds good @demiankatz. Thanks for the help and suggestions for improving the code. I really appreciate it. I'll get to work on refactoring and simplifying the code per your suggestions.

lmgonzales commented:
@demiankatz , Can you help me with this part of the refactoring? I've not used HashMaps much before, but what I'm trying to do is check if the coordinate set is a point coordinate, and if so fill the empty character, string values with the corresponding data from the d/e or f/g character, string pair. I'm getting this error: 

> no suitable method found for put(char,java.lang.Object)
>                 full_coords.put('e', coords.get('d'));
>                            ^
>     method java.util.Map.put(java.lang.Character,java.lang.String) is not applicable
>       (argument mismatch; java.lang.Object cannot be converted to java.lang.String)
>     method java.util.AbstractMap.put(java.lang.Character,java.lang.String) is not applicable
>       (argument mismatch; java.lang.Object cannot be converted to java.lang.String)
>     method java.util.HashMap.put(java.lang.Character,java.lang.String) is not applicable
>       (argument mismatch; java.lang.Object cannot be converted to java.lang.String)
> 
> 

Here's the code I'm working on. 
```
    /**
    * If coordinates are a point, fill empty N/S or E/W coordinate
    *
    * @param  HashMap coords
    * @return HashMap full_coords
    */
   protected HashMap<Character, String> fillEmptyPointCoordinates(HashMap coords) {
        HashMap<Character, String> full_coords = coords;
        if (coords.containsKey('d') && !coords.containsKey('e') && coords.containsKey('f') && !coords.containsKey('g')) {
                full_coords.put('e', coords.get('d'));
                full_coords.put('g', coords.get('f'));
        }
        if (coords.containsKey('e') && !coords.containsKey('d') && coords.containsKey('g') && !coords.containsKey('h')) {
                full_coords.put('d', coords.get('e'));
                full_coords.put('f', coords.get('g'));
        }
        return full_coords;
   }
```
Any ideas on what to do?  Thanks!

lmgonzales commented:
Figured it out. I had to cast the `coords.get('d')` to `coords.get('d').toString()`

lmgonzales commented:
@demiankatz , 
Thanks again for your suggestions on restructuring the code. I've completed the refactoring of the code, and it's much more concise now. Please let me know if you see any other places for improvement.

demiankatz commented:
@lmgonzales, thanks for all the work on this (and for figuring out your problem on your own -- sorry I was too slow there). It's looking great to me. I'll take a few minutes to test it again in the next few days, and then it should be ready to merge.

lmgonzales commented:
Thanks @demiankatz! Sounds good.

demiankatz commented:
I've forward-ported your fixes to release 4.0 here: https://github.com/vufind-org/vufind/commit/a81ef438d5ef26b2d0dd6c98661e6ef30db75ea8 -- I think all is well, but please let me know if you run into problems whenever you get around to upgrading to the latest master code.

lmgonzales commented:
Thank you very much Demian. I’ll let you know if I run into any problems
when we upgrade to 4.0. Thanks again!



*From:* Demian Katz [mailto:notifications@github.com]
*Sent:* Thursday, May 25, 2017 12:31 PM
*To:* vufind-org/vufind
*Cc:* Leila Gonzales; Mention
*Subject:* Re: [vufind-org/vufind] bug fix for geo indexing and display
(#964)



I've forward-ported your fixes to release 4.0 here: a81ef43
<https://github.com/vufind-org/vufind/commit/a81ef438d5ef26b2d0dd6c98661e6ef30db75ea8>
-- I think all is well, but please let me know if you run into problems
whenever you get around to upgrading to the latest master code.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/vufind-org/vufind/pull/964#issuecomment-304101819>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AMc0UW0skj4qGn8avy3Latv7zNXrlIX7ks5r9dcCgaJpZM4NV6cF>
.
